core/src/labels/labelCollider.cpp

```diff
@@ -0,0 +1,221 @@
+    m_screenSize = glm::vec2{ _tileSize * m_tileScale };
+}
+
+void LabelCollider::addLabels(std::vector<std::unique_ptr<Label>>& _labels) {
+
+    for (auto& label : _labels) {
+        if (label->canOcclude()) {
+            m_labels.push_back(label.get());
+        }
+    }
+}
+
+void LabelCollider::handleRepeatGroup(size_t startPos) {
+
+    float threshold2 = pow(m_labels[startPos]->options().repeatDistance, 2);
+    size_t repeatGroup = m_labels[startPos]->options().repeatGroup;
+
+    // Get the range
+    size_t endPos = startPos;
+    for (;startPos > 0; startPos--) {
+        if (m_labels[startPos-1]->options().repeatGroup != repeatGroup) { break; }
+    }
+    for (;endPos < m_labels.size()-1; endPos++) {
+        if (m_labels[endPos+1]->options().repeatGroup != repeatGroup) { break; }
+    }
+
+
+    for (size_t i = startPos; i < endPos; i++) {
+        Label* l1 = m_labels[i];
+        if (l1->isOccluded()) { continue; }
+
+        for (size_t j = i+1; j <= endPos; j++) {
+            Label* l2 = m_labels[j];
+            if (l2->isOccluded()) { continue; }
+
+            float d2 = distance2(l1->center(), l2->center());
+            if (d2 < threshold2) {
+                l2->occlude();
+            }
+        }
+    }
+}
+
+void LabelCollider::process() {
+
+    // Sort labels so that all labels of one repeat group are next to each other
+    std::sort(m_labels.begin(), m_labels.end(),
+              [](auto* l1, auto* l2) {
+                  if (l1->options().priority != l2->options().priority) {
+                      // lower numeric priority means higher priority
+                      return l1->options().priority > l2->options().priority;
+                  }
+                  if (l1->options().repeatGroup != l2->options().repeatGroup) {
+                      return l1->options().repeatGroup < l2->options().repeatGroup;
+                  }
+
+                  if (l1->type() == Label::Type::line && l2->type() == Label::Type::line) {
+                      // Prefer the label with longer line segment as it has a chance
+                      // to be shown earlier (also on the lower zoom-level)
+                      // TODO compare fraction segment_length/label_width
+                      return glm::length(l1->transform().modelPosition1 - l1->transform().modelPosition2) >
+                          glm::length(l2->transform().modelPosition1 - l2->transform().modelPosition2);
+                  }
+
+                  return l1->hash() < l2->hash();
+              });
+
+    // Project tile to NDC (-1 to 1, y-up)
+    glm::mat4 mvp{1};
+    // Scale tile to 'fullscreen'
+    mvp[0][0] = 2;
+    mvp[1][1] = -2;
+    // Place tile centered
+    mvp[3][0] = -1;
+    mvp[3][1] = 1;
+
+    for (auto* label : m_labels) {
+        label->update(mvp, m_screenSize, 1, true);
+
+         m_aabbs.push_back(label->aabb());
+    }
+
+    m_isect2d.resize({m_screenSize.x / 128, m_screenSize.y / 128}, m_screenSize);
```

Comment:
Hard to say what a good value is for the grid :)
With the default pixel scale a tile is projected to 512x512, so this would make a 4x4 grid which seem reasonable.